Let flake8 handle the default config files

OmeGak opened the pull request:
Flake8 is capable of correctly loading the config from coexisting default files. Explicitly passing files from linter-flake8 overrides this behavior.

Fixes #554.

Arcanemagus commented on package.json:
```diff
@@ -25,12 +25,8 @@
-        "setup.cfg"
-      ],
-      "description": "flake config file relative path from project (e.g. tox.ini or .flake8rc)",
+      "default": [],
```
This setting is used to determine whether or not a configuration file exists for the project, which then controls whether the overrides are taken into account or not. So making it blank like this entirely breaks that functionality.

A better way to handle this would be to have this setting be an _addition_ to the list of configuration files, with the default list `flake8` searches for stored in the code. Use the merged list to see if there is a config file at all, and use the custom list to determine if `--config` needs to be specified.

lucasdf replied:
Doesn't Flake8 already find and use the appropriated configuration file? As it is now `--config` always override the default behavior due to the default values. Check the discussion on #554

Arcanemagus replied:
With the current state of this PR [this condition](https://github.com/AtomLinter/linter-flake8/blob/0f3d35965a0761c6cb229ebc7a1d3938215376a2/lib/main.js#L212) will virtually always evaluate to `false`, meaning all of the settings like [these](https://github.com/AtomLinter/linter-flake8/blob/0f3d35965a0761c6cb229ebc7a1d3938215376a2/package.json#L38-L65) that add in override flags like `--max-complexity` will come into play, which would stomp all over any configuration file `flake8` uses.

The detection of the configuration file is what needs to be fixed _first_, then once the configuration file is being properly detected we can decide whether or not to force it with `--config`. I was proposing a super basic way of if one of the default config files is found (`.flake8`, `tox.ini`...) to not pass `--config` and _also_ not pass things like `--max-complexity`. If the user specifies a custom file in the settings and it is detected then that would always be sent in via `--config`.

An extension to that, which sounds like it is necessary for files like `tox.ini` is to determine if they have a `flake8` section, and if not then allow the overrides like `--max-complexity` through.

OmeGak replied:
I'm fine with keeping this PR on hold and I may take a stab at detecting default config files. Do you want a separate issue for that?

Arcanemagus replied:
Up to you, I'm fine with this being the "fix all the config things" PR 😛.

OmeGak commented:
Bump. This PR now:

- Lets `flake8` handle config files unless one is manually specified.
- If the config file is either manually specified or found among the `flake8` default config files (`.flake8`, `tox.ini`, `setup.cfg`), flags don't take effect (`--max-complexity`, etc.).

Downside, this logic is not apparent from reading the settings descriptions, so I'm going to try and see if I can improve that a bit as well.

--- 

I updated the PR following @Arcanemagus' suggestion:

> I was proposing a super basic way of if one of the default config files is found (`.flake8`, `tox.ini`...) to not pass `--config` and also not pass things like `--max-complexity`. If the user specifies a custom file in the settings and it is detected then that would always be sent in via `--config`.

I would also like to mention:

> The detection of the configuration file is what needs to be fixed first, then once the configuration file is being properly detected we can decide whether or not to force it with `--config`.

`flake8` manages to determine the correct config file if no `--config` is passed.

OmeGak commented:
Updated again. Hopefully this will make the obscure logic a bit less frustrating.

- Improved organization of config schema and setting descriptions.
- Added new checkbox to override settings from config files.

---

<img width="747" alt="screen shot 2018-03-19 at 00 13 31" src="https://user-images.githubusercontent.com/716307/37572399-beed62ea-2b0a-11e8-87c2-3aef8843fb0e.png">

OmeGak commented:
Ping.

OmeGak commented:
Ping @Arcanemagus. What is the blocker for this PR?

Arcanemagus commented on lib/main.js:
```diff
@@ -26,6 +26,11 @@ function loadDeps() {
 // Local variables
 const parseRegex = /(\d+):(\d+):\s(([A-Z])\d{2,3})\s+(.*)/g;
 const execPathVersions = new Map();
+const defaultConfigFiles = [
```
Is there a reason this is statically declared here, instead of setting this as the default value for the `configFile` option? Should we care if the config file found is a default one Flake8 handles, or one specified by the user?

OmeGak replied:
If we set these files as the default value of `configFile` then we also need to maintain the order in which `flake8` is supposed to be looking for them. This seemed to be something that changed already in the past, as you mentioned in https://github.com/AtomLinter/linter-flake8/issues/554#issuecomment-353121371.

Instead, calling `flake8` without the `--config` parameter makes it choose the default config file in the correct order. I believe this is a more robust approach than us hardcoding the order those default config files need to be chosen.

OmeGak commented:
> It looks like you have changed the config schema since I last looked at this, but haven't put in any method for migrating any user's existing configuration to the new one, which currently makes this a breaking change. Can you add some functionality that accomplishes this?

Migrating already existing user configuration to a new schema seems like something that should be pretty common in the Atom plugin ecosystem, but I cannot find any recommendation on how to go about it. Am I missing something obvious before I end up coding something overcomplicated?

Arcanemagus commented:
You could go anywhere between what we're [doing in `linter-eslint`](https://github.com/AtomLinter/linter-eslint/blob/28960f255c54a80f245571d490520247061f59a7/src/migrate-config-options.js) to just a block that checks for the old values and migrates them.

With 7 values to migrate it's tough to say which I'd go with, but _probably_ something closer to what is being done in `linter-eslint`.

OmeGak commented:
Thanks @Arcanemagus. I did nothing fancy in the end, just adjusting the mechanism from linter-eslint. It pains me to see so much code repetition between plugins that could share it, but it does the job for now.

Scotchester commented:
Any particular reason why this was closed? I just ran into the issue that this PR purportedly fixes (#554).

OmeGak commented:
Hey @Scotchester, I closed this PR because I'm not using Atom anymore and I'm not planning to finish the development. The code was working, but I don't know what was preventing @Arcanemagus merging it. If you want to follow up, please feel free to fork my fork and open a new PR. 🙏
